plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/CallSiteInline.hs

```diff
@@ -60,120 +64,110 @@ becomes
-    => Term tyname name uni fun ann -- ^ The `body` of the `Let` term.
-    -> InlineM tyname name uni fun ann (Term tyname name uni fun ann)
-considerInlineSat tm = do
-    -- collect all the arguments of the term being applied to
-    case splitApplication tm of
-      -- if it is a `Var` that is being applied to, check to see if it's fully applied
-      (Var _ann name, ctx) -> do
-        maybeVarInfo <- gets (lookupVarInfo name)
-        case maybeVarInfo of
-          Just varInfo -> do
-            let body = varBody varInfo
-                defAsInlineTerm = varDef varInfo
-                inlineTermToTerm :: InlineTerm tyname name uni fun ann
-                  -> Term tyname name uni fun ann
-                inlineTermToTerm (Done (Dupable var)) = var
-                def = inlineTermToTerm defAsInlineTerm
-                fullyApplied = isFullyApplied (arity varInfo) ctx
-                -- It is the body that we will be left with in the program after we have
-                -- reduced the saturated application, so the size increase we will be left
-                -- with comes from the body, and that is what we need to check is okay
-                bodySizeOk = sizeIsAcceptable body
+-- | Fully apply the RHS of the given variable to the given arguments, and beta-reduce
+-- the application, if possible.
+fullyApplyAndBetaReduce ::
+  forall tyname name uni fun ann.
+  (InliningConstraints tyname name uni fun) =>
+  -- | The variable
+  VarInfo tyname name uni fun ann ->
+  -- | The arguments
+  AppContext tyname name uni fun ann ->
+  InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+fullyApplyAndBetaReduce info = go (varBody info) (varArity info)
+  where
+    go ::
+      Term tyname name uni fun ann ->
+      Arity tyname name ->
+      AppContext tyname name uni fun ann ->
+      InlineM tyname name uni fun ann (Maybe (Term tyname name uni fun ann))
+    go acc arity args = case (arity, args) of
+      -- success
+      ([], _) -> pure . Just $ fillAppContext acc args
+      (TermParam param : arity', TermAppContext arg _ args') ->
+        ifM
+          (safeToSubst param arg)
+          ( go
+              (termSubstNames (\n -> if n == param then Just arg else Nothing) acc)
+              arity'
+              args'
+          )
+          (pure Nothing)
+      (TypeParam param : arity', TypeAppContext arg _ args') ->
+        go
+          (termSubstTyNames (\n -> if n == param then Just arg else Nothing) acc)
+          arity'
+          args'
+      _ -> pure Nothing
+
+    -- Checks whether it is safe (i.e., preserves semantics) to perform beta reduction,
+    -- namely, turning `(\a -> body) arg` into `body [a := arg]`.
+    safeToSubst :: name -> Term tyname name uni fun ann -> InlineM tyname name uni fun ann Bool
+    safeToSubst n = effectSafe (varBody info) Strict n <=< checkPurity
+
+-- | Consider whether to inline an application.
+inlineSaturatedApp ::
```

Comment:
I like the old name `considerInlineSat` better, it shows that it may or may not be inlined. Perhaps it should be changed to `considerInlineBetaReduceSatFn` or something like that.

Reply:
I don't really care that much, but this is in line with GHC's naming actually.